src/IO/WriteBufferFromPocoSocketChunked.cpp

```diff
@@ -0,0 +1,207 @@
+#include <IO/WriteBufferFromPocoSocketChunked.h>
+#include <Common/logger_useful.h>
+#include <IO/NetUtils.h>
+
+
+namespace
+{
+
+template <typename T>
+const T & setValue(T * typed_ptr, std::type_identity_t<T> val)
+{
+    memcpy(typed_ptr, &val, sizeof(T));
+    return *typed_ptr;
+}
+
+}
+
+namespace DB
+{
+
+void WriteBufferFromPocoSocketChunked::enableChunked()
+{
+    chunked = true;
+    /// Initialize next chunk
+    chunk_size_ptr = reinterpret_cast<decltype(chunk_size_ptr)>(pos);
+    pos += std::min(available(), sizeof(*chunk_size_ptr));
+}
+
+void WriteBufferFromPocoSocketChunked::finishChunk()
+{
+    if (!chunked)
+        return;
+
+    if (pos <= reinterpret_cast<Position>(chunk_size_ptr) + sizeof(*chunk_size_ptr))
+    {
+        if (chunk_size_ptr == last_finish_chunk) // prevent duplicate finish chunk
+            return;
+
+        /// If current chunk is empty it means we are finishing a chunk previously sent by next(),
+        /// we want to convert current chunk header into end-of-chunk marker and initialize next chunk.
+        /// We don't need to worry about if it's the end of the buffer because next() always sends the whole buffer
+        /// so it should be a beginning of the buffer.
+
+        chassert(reinterpret_cast<Position>(chunk_size_ptr) == working_buffer.begin());
```

Comment:
It may also happen that `finishChunk` was called right after `enableChunk`. In this case `chunk_size_ptr` won't be the beginning of the buffer. This should not happen actually as we don't support sending emopty chunks, but it will be better to check this case and throw proper exception message

Reply:
it should be prevented by previous `if` - it's already working for two consecutive finishChunk - let's fix it for after `enableChunk` case by initializing `last_finish_chunk` in the `enableChunk`...